src/Block/AdminStatsBlockService.php

```diff
@@ -48,14 +48,14 @@ public function execute(BlockContextInterface $blockContext, Response $response
 
         $filters = $blockContext->getSetting('filters');
 
-        if (!isset($filters['_per_page'])) {
-            $filters['_per_page'] = ['value' => $blockContext->getSetting('limit')];
-        }
-
         foreach ($filters as $name => $data) {
             $datagrid->setValue($name, $data['type'] ?? null, $data['value']);
         }
 
+        if (!isset($filters['_per_page'])) {
+            $filters['_per_page'] = ['value' => $blockContext->getSetting('limit')];
```

Comment:
This change seems to have no sense, `$filters` array is only used in the previous `foreach` loop, so changing its value here has no effect.

Reply:
Ok ! This change works only because there are no values for `_per_page` set to `Datagrid`... 
I tried to fix the issue in Datagrid but the hidden field where _per_page is, does not want the array for value and I could not call a transformer...

Reply:
me too ;) do you know how to build a filter?
$datagrid->addFilter(FilterInterface ...)

Reply:
The problem with this case is in datagrid and not here. If we look at how Admin Crud manages filters, we see that it does not use the $dataGrid->addValue for the _per_page parameter. The Admin Crud pass it directly to the Pager.
This is because when you add it through the addValue function, all fields are validated in the form and the _per_page field is a Hidden type that requires a number or text. The management of the array with the key "value" is done after the validation when it should be done before in the case of this field. 

This is my opinion ;-)

Reply:
@GCalmels, I send to you my fix proposal. Please, review it and if you are agree, merge the change. In my environment, the change It's working fine.

Reply:
@phansys what do you think about the @MAGlCUS proposal ? It's resolve the issue for me but omits the `type`...

Reply:
This seems like a workaround.

Reply:
@GCalmels, @phansys, you have to analyse the function buildPager in Datagrid.php. In this function, where the pager is builded, we have:
```
foreach ($this->getFilters() as $name => $filter) {
            list($type, $options) = $filter->getRenderSettings();

            $this->formBuilder->add($filter->getFormName(), $type, $options);
        }
```
First, all filters are added to the form. After that, It is added _sort_by, _sort_order, _page and _per_page with a hiddenType:
```
$hiddenType = HiddenType::class;

        $this->formBuilder->add('_sort_by', $hiddenType);
        $this->formBuilder->get('_sort_by')->addViewTransformer(new CallbackTransformer(
            static function ($value) {
                return $value;
            },
            static function ($value) {
                return $value instanceof FieldDescriptionInterface ? $value->getName() : $value;
            }
        ));

        $this->formBuilder->add('_sort_order', $hiddenType);
        $this->formBuilder->add('_page', $hiddenType);
        $this->formBuilder->add('_per_page', $hiddenType);
```
When we have added all the fields, It submitted the values:
```
$this->form = $this->formBuilder->getForm();
$this->form->submit($this->values);
```
In the other hand, as you know, the function setValue make this:
```
$this->values[$name] = [
                'type' => $operator,
                'value' => $value,
            ];
```
If we use setValue to set the values for "_page" or "_per_page", we'll obtain the error in form validation when we submit the form. "_page" and "_per_page" have a hiddenType and we pass to the "submit" function an array of "type" and "value".

With this scenario, I think we have 2 options, we can make a new function to setPerPage and add to the array values with the correct value or modify setValue to take into account the "special" values for "_per_page", "pager" and if we wants for "sort_order" and "sort_by", although these last values are a little bit different.

I hope my explanation was precise enough.